kernel/src/io_apic.c

```diff
@@ -0,0 +1,83 @@
+#include "io_apic.h"
+
+#include "port_io.h"
+
+// Gets lower register offset for irq redtable entry
+#define REDTBL_OFFSET(irq) (2 * irq + 0x10)
+#define GET_BITRANGE_VALUE(num, lb, ub) ((num & ((1U << ub) - (1U << lb))) >> lb)
+
+uint32_t read_ioapic_register(void* ioapic_address, uint8_t offset) {
+    // tell IOREDSEL where to read
+    *(volatile uint32_t*)(ioapic_address) = offset;
+
+    // return IOREGWIN
+    return *(volatile uint32_t*)(ioapic_address + 0x10);
+}
+
+void write_ioapic_register(void* ioapic_address, uint8_t offset, uint32_t value) {
+    // tell IOREDSEL where to write
+    *(volatile uint32_t*)(ioapic_address) = offset;
+
+    // write to IOREGWIN
+    *(volatile uint32_t*)(ioapic_address + 0x10) = value;
+}
+
+uint32_t read_ioapic_id(void* ioapic_address) {
```

Comment:
Why not use a packed struct to read values of the IO APIC instead of a bunch of functions?